api/valor_api/backend/core/llm_clients.py

```diff
@@ -0,0 +1,521 @@
+        statements: list[str]
+            The statements to evaluate the validity of.
+
+        Returns
+        -------
+        list[dict[str,str]]
+            The list of verdicts for each statement. Each verdict is a dictionary with the "verdict" and optionally a "reason".
+        """
+        messages = [
+            {"role": "system", "content": DEFAULT_SYSTEM_PROMPT},
+            {
+                "role": "user",
+                "content": _generate_verdicts_instruction(query, statements),
+            },
+        ]
+
+        response = self(messages)
+
+        try:
+            response = trim_and_load_json(response)
+            assert type(response) == dict
+            assert "verdicts" in response
+            verdicts = response["verdicts"]
+            assert type(verdicts) == list
+            assert len(verdicts) == len(statements)
+            assert all(
+                verdict["verdict"] in ["yes", "no", "idk"]
+                for verdict in verdicts
+            )
+        except Exception:
+            raise ValueError(
+                f"LLM response was not a valid list of verdicts: {response}"
+            )
+
+        return verdicts
+
+    def answer_relevance(
+        self,
+        query: str,
+        text: str,
+    ) -> float:
+        """
+        Compute answer relevance, the proportion of statements that are relevant to the query, for a single piece of text.
+
+        Parameters
+        ----------
+        query: str
+            The query to evaluate the statements against.
+        text: str
+            The text to extract statements from.
+
+        Returns
+        -------
+        float
+            The answer relevance score will be evaluated as a float between 0 and 1, with 1 indicating that all statements are relevant to the query.
+        """
+        statements = self._generate_statements(text)
+        verdicts = self._generate_verdicts(query, statements)
+        return sum(
+            1 for verdict in verdicts if verdict["verdict"] == "yes"
+        ) / len(verdicts)
+
+    def coherence(
+        self,
+        text: str,
+    ) -> int:
+        """
+        Compute coherence, the collective quality of all sentences, for a single piece of text.
+
+        Parameters
+        ----------
+        text: str
+            The text to be evaluated.
+
+        Returns
+        -------
+        int
+            The coherence score will be evaluated as an integer, with 1 indicating the lowest coherence and 5 the highest coherence.
+        """
+        messages = [
+            {"role": "system", "content": DEFAULT_SYSTEM_PROMPT},
+            {"role": "user", "content": _get_coherence_instruction(text)},
+        ]
+
+        response = self(messages)
+
+        try:
+            # Valid responses: "5", "\n5", "5\n", "5.", " 5", "5 {explanation}", etc.
+            ret = int(response.strip()[0])
+            assert ret in {1, 2, 3, 4, 5}
+        except Exception:
```

Comment:
same comment as for answer relevance above. This might make it hard for the user to understand the issue with the llm's response